core/services/keeper/orm.go

```diff
@@ -99,25 +110,45 @@ DELETE FROM upkeep_registrations WHERE registry_id IN (
 	return rowsAffected, nil
 }
 
-func (korm ORM) EligibleUpkeepsForRegistry(registryAddress ethkey.EIP55Address, blockNumber, gracePeriod int64) (upkeeps []UpkeepRegistration, err error) {
+//EligibleUpkeepsForRegistry fetches eligible upkeeps for processing
+//The query checks the following conditions
+// - checks the registry address is correct and the registry has some keepers associated
+// -- is it my turn AND my keeper was not the last perform for this upkeep OR my keeper was the last before BUT it is past the grace period
+// -- OR is it my buddy's turn AND they were the last keeper to do the perform for this upkeep
+func (korm ORM) EligibleUpkeepsForRegistry(registryAddress ethkey.EIP55Address, head *types.Head, gracePeriod int64) (upkeeps []UpkeepRegistration, err error) {
+	registry, err := korm.RegistryByContractAddress(registryAddress)
+	if err != nil {
+		return nil, errors.Wrap(err, "EligibleUpkeepsForRegistry failed to get a registry by address")
+	}
+	blockNumber := head.Number
+	binaryHash := binaryOfFirstHashInTurn(blockNumber, registry, head)
+
 	stmt := `
 SELECT upkeep_registrations.* FROM upkeep_registrations
 INNER JOIN keeper_registries ON keeper_registries.id = upkeep_registrations.registry_id
 WHERE
 	keeper_registries.contract_address = $1 AND
-	keeper_registries.num_keepers > 0 AND
-	(
-		upkeep_registrations.last_run_block_height = 0 OR (
-			upkeep_registrations.last_run_block_height + $2 < $3 AND
-			upkeep_registrations.last_run_block_height < ($3 - ($3 % keeper_registries.block_count_per_turn))
-		)
-	) AND
-	keeper_registries.keeper_index = (
-		upkeep_registrations.positioning_constant + (($3 - ($3 % keeper_registries.block_count_per_turn)) / keeper_registries.block_count_per_turn)
-	) % keeper_registries.num_keepers
-ORDER BY upkeep_registrations.id ASC, upkeep_registrations.upkeep_id ASC
+	keeper_registries.num_keepers > 0 AND 
+    ((
+                keeper_registries.keeper_index = ((CAST(upkeep_registrations.upkeep_id AS bit(32)) #
```

Comment:
Not something we need to change now, but just calling out for future:

UpkeepIDs are `uint256` and starting on keeper registry 1.2 are non-sequential, meaning they won't fit in 32 bits. We'll need to update this for the next version of the registry.

Reply:
should be as easy as casting everything to a higher bit.